File: grooves-example-javaee/src/main/java/grooves/example/javaee/BootStrap.java
package grooves.example.javaee;

import static rx.RxReactiveStreams.toObservable;

import com.github.rahulsom.grooves.api.snapshots.Snapshot;
import com.github.rahulsom.grooves.test.EventsDsl;
import com.github.rahulsom.grooves.test.OnSpec;
import grooves.example.javaee.domain.*;
import grooves.example.javaee.queries.PatientAccountQuery;
import grooves.example.javaee.queries.PatientHealthQuery;
import java.math.BigDecimal;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Supplier;
import javax.annotation.PostConstruct;
import javax.ejb.Singleton;
import javax.ejb.Startup;
import javax.inject.Inject;

/**
 * Initializes data for app.
 */
@SuppressWarnings("UnusedReturnValue")
@Singleton
@Startup
public class BootStrap {

    private static final String ANNUAL_PHYSICAL_NAME = "ANNUALPHYSICAL";
    private static final String ANNUAL_PHYSICAL_COST = "170.00";
    private static final String FLU_SHOT_NAME = "FLUSHOT";
    private static final String FLU_SHOT_COST = "32.40";
    private static final String GLUCOSE_TEST_NAME = "GLUCOSETEST";
    private static final String GLUCOSE_TEST_COST = "78.93";

    private static long idGenerator = 1;

    private Date currDate = date("2016-01-01");
    private PatientAccountQuery patientAccountQuery;
    private PatientHealthQuery patientHealthQuery;
    private Database database;

    /**
     * Injects a {@link PatientAccountQuery}.
     *
     * @param patientAccountQuery The {@link PatientAccountQuery} to inject.
     */
    @Inject
    public void setPatientAccountQuery(PatientAccountQuery patientAccountQuery) {
        this.patientAccountQuery = patientAccountQuery;
    }

    /**
     * Injects a {@link PatientHealthQuery}.
     *
     * @param patientHealthQuery The {@link PatientHealthQuery} to inject.
     */
    @Inject
    public void setPatientHealthQuery(PatientHealthQuery patientHealthQuery) {
        this.patientHealthQuery = patientHealthQuery;
    }

    /**
     * Sets a database instance.
     *
     * @param database the Database to connect to
     */
    @Inject
    public void setDatabase(Database database) {
        this.database = database;
    }

    /**
     * Sets up data for tests.
     */
    @PostConstruct
    public void init() {
        setupJohnLennon();
        setupRingoStarr();
        setupPaulMcCartney();
        setupFreddieMercury();
        setupTinaFeyAndSarahPalin();
    }

    private Patient save(Patient patient) {
        patient.setId(database.patients().count() + 1L);
        database.addPatient(patient);
        return patient;
    }

    private Patient setupJohnLennon() {

        Patient patient = save(new Patient("42"));

        return on(patient, it -> {
            it.apply(new PatientCreated("John Lennon"));
            it.apply(new ProcedurePerformed(FLU_SHOT_NAME, new BigDecimal(FLU_SHOT_COST)));
            it.apply(new ProcedurePerformed(GLUCOSE_TEST_NAME, new BigDecimal(GLUCOSE_TEST_COST)));
            it.apply(new PaymentMade(new BigDecimal("100.25")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);

            it.apply(new ProcedurePerformed(ANNUAL_PHYSICAL_NAME, new BigDecimal(ANNUAL_PHYSICAL_COST)));
            it.apply(new PaymentMade(new BigDecimal("180.00")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });
    }

    private Patient setupRingoStarr() {
        Patient patient = save(new Patient("43"));

        return on(patient, it -> {
            it.apply(new PatientCreated("Ringo Starr"));
            it.apply(new ProcedurePerformed(ANNUAL_PHYSICAL_NAME, new BigDecimal(ANNUAL_PHYSICAL_COST)));
            it.apply(new ProcedurePerformed(GLUCOSE_TEST_NAME, new BigDecimal(GLUCOSE_TEST_COST)));
            it.apply(new PaymentMade(new BigDecimal("100.25")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);

            it.apply(new ProcedurePerformed(FLU_SHOT_NAME, new BigDecimal(FLU_SHOT_COST)));
            it.apply(new PaymentMade(new BigDecimal("180.00")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });
    }

    private Patient setupPaulMcCartney() {
        Patient patient = save(new Patient("44"));

        return on(patient, it -> {
            it.apply(new PatientCreated("Paul McCartney"));
            it.apply(new ProcedurePerformed(ANNUAL_PHYSICAL_NAME, new BigDecimal(ANNUAL_PHYSICAL_COST)));
            ProcedurePerformed gluc = (ProcedurePerformed)
                    it.apply(new ProcedurePerformed(GLUCOSE_TEST_NAME, new BigDecimal(GLUCOSE_TEST_COST)));
            it.apply(new PaymentMade(new BigDecimal("100.25")));
            it.apply(new PatientEventReverted(gluc.getId()));
            PaymentMade pmt = (PaymentMade) it.apply(new PaymentMade(new BigDecimal("30.00")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);

            it.apply(new PatientEventReverted(pmt.getId()));
            it.apply(new PaymentMade(new BigDecimal("60.00")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);

            it.apply(new PaymentMade(new BigDecimal("60.00")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });
    }

    private Patient setupFreddieMercury() {
        Patient patient = save(new Patient("45"));
        Patient patient2 = save(new Patient("46"));

        on(patient, it -> {
            it.apply(new PatientCreated("Farrokh Bulsara"));
            it.apply(new ProcedurePerformed(ANNUAL_PHYSICAL_NAME, new BigDecimal(ANNUAL_PHYSICAL_COST)));
            it.apply(new ProcedurePerformed(GLUCOSE_TEST_NAME, new BigDecimal(GLUCOSE_TEST_COST)));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });

        on(patient2, it -> {
            final String name = "Freddie Mercury";
            it.apply(new PatientCreated(name));
            it.apply(new PaymentMade(new BigDecimal("100.25")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });

        final Calendar calendar = Calendar.getInstance();
        calendar.setTime(currDate);
        calendar.add(Calendar.DAY_OF_MONTH, 1);
        currDate = calendar.getTime();
        merge(patient, patient2);
        return patient;
    }

    private Patient setupTinaFeyAndSarahPalin() {
        Patient patient = save(new Patient("47"));
        Patient patient2 = save(new Patient("48"));

        on(patient, it -> {
            it.apply(new PatientCreated("Tina Fey"));
            it.apply(new ProcedurePerformed(ANNUAL_PHYSICAL_NAME, new BigDecimal(ANNUAL_PHYSICAL_COST)));
            it.apply(new ProcedurePerformed(GLUCOSE_TEST_NAME, new BigDecimal(GLUCOSE_TEST_COST)));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });

        on(patient2, it -> {
            final String name = "Sarah Palin";
            it.apply(new PatientCreated(name));
            it.apply(new PaymentMade(new BigDecimal("100.25")));

            it.snapshotWith(patientAccountQuery);
            it.snapshotWith(patientHealthQuery);
        });

        final Calendar calendar = Calendar.getInstance();
        calendar.setTime(currDate);
        calendar.add(Calendar.DAY_OF_MONTH, 1);
        currDate = calendar.getTime();
        final PatientDeprecatedBy mergeEvent = merge(patient, patient2);

        on(patient, it -> it.apply(new PatientEventReverted(mergeEvent.getId())));
        on(
                patient2,
                it -> it.apply(new PatientEventReverted(mergeEvent.getConverse().getId())));

        return patient;
    }

    /**
     * Merges `self` into `into`.
     *
     * @param self The aggregate to be deprecated
     * @param into The aggregate to survive
     * @return The DeprecatedBy event
     */
    private PatientDeprecatedBy merge(Patient self, Patient into) {
        PatientDeprecatedBy e1 = new PatientDeprecatedBy(into);

        e1.setAggregate(self);
        e1.setTimestamp(currDate);
        e1.setPosition(database.events()
                        .filter(x -> Objects.equals(x.getAggregate(), self))
                        .count()
                + 1);
        e1.setId(database.events().count() + 1);

        PatientDeprecates e2 = new PatientDeprecates(e1, self);

        e2.setAggregate(into);
        e2.setTimestamp(currDate);
        e2.setPosition(database.events()
                        .filter(x -> Objects.equals(x.getAggregate(), into))
                        .count()
                + 1);
        e2.setId(database.events().count() + 2);

        e1.setConverse(e2);

        database.addEvent(e1);
        database.addEvent(e2);
        return toObservable(e2.getConverseObservable()).toBlocking().single();
    }

    private Date date(String yyyyMMdd) {
        try {
            return new SimpleDateFormat("yyyy-MM-dd").parse(yyyyMMdd);
        } catch (ParseException ignore) {
            return null;
        }
    }

    private Patient on(Patient patient, Consumer<OnSpec> closure) {
        final Consumer<?> eventSaver = object -> {
            if (object instanceof PatientEvent patientEvent) {
                patientEvent.setId(idGenerator++);
                database.addEvent((PatientEvent) object);
            } else if (object instanceof Snapshot snapshot) {
                snapshot.setId(idGenerator++);
                database.addSnapshot(object);
            }
        };

        Supplier<Long> positionSupplier = () -> database.events()
                        .filter(x -> toObservable(x.getAggregateObservable())
                                .toBlocking()
                                .single()
                                .equals(patient))
                        .count()
                + 1;

        Supplier<Date> dateSupplier = () -> {
            final Calendar calendar = Calendar.getInstance();
            calendar.setTime(currDate);
            calendar.add(Calendar.DAY_OF_MONTH, 1);
            currDate = calendar.getTime();
            return currDate;
        };

        EventsDsl<Patient, Long, PatientEvent> dsl = new EventsDsl<>();
        return dsl.on(patient, eventSaver, positionSupplier, dateSupplier, closure::accept);
    }
}

File: grooves-example-springboot-kotlin/src/main/kotlin/grooves/boot/kotlin/BootStrap.kt
package grooves.boot.kotlin

import com.github.rahulsom.grooves.api.snapshots.Snapshot
import com.github.rahulsom.grooves.queries.QuerySupport
import com.github.rahulsom.grooves.test.OnSpec
import grooves.boot.kotlin.domain.Patient
import grooves.boot.kotlin.domain.PatientEvent
import grooves.boot.kotlin.domain.PatientEvent.Applicable.Created
import grooves.boot.kotlin.domain.PatientEvent.Applicable.PaymentMade
import grooves.boot.kotlin.domain.PatientEvent.Applicable.ProcedurePerformed
import grooves.boot.kotlin.queries.PatientAccountQuery
import grooves.boot.kotlin.queries.PatientHealthQuery
import grooves.boot.kotlin.repositories.PatientAccountRepository
import grooves.boot.kotlin.repositories.PatientBlockingRepository
import grooves.boot.kotlin.repositories.PatientEventBlockingRepository
import grooves.boot.kotlin.repositories.PatientHealthRepository
import jakarta.annotation.PostConstruct
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.repository.reactive.ReactiveCrudRepository
import org.springframework.stereotype.Component
import java.math.BigDecimal
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.TimeZone

@Component
class BootStrap constructor(
    @param:Autowired val patientRepository: PatientBlockingRepository,
    @param:Autowired val patientEventRepository: PatientEventBlockingRepository,
    @param:Autowired val patientAccountQuery: PatientAccountQuery,
    @param:Autowired val patientHealthQuery: PatientHealthQuery,
    @param:Autowired val patientAccountRepository: PatientAccountRepository,
    @param:Autowired val patientHealthRepository: PatientHealthRepository,
) {
    @PostConstruct
    fun init() {
        setupJohnLennon()
        setupRingoStarr()
        setupPaulMcCartney()
        setupFreddieMercury()
        setupTinaFeyAndSarahPalin()
    }

    fun setupJohnLennon() {
        on(patientRepository.save(Patient("42"))) {
            it.apply(Created("John Lennon"))
            it.apply(ProcedurePerformed("FLUSHOT", BigDecimal("32.40")))
            it.apply(ProcedurePerformed("GLUCOSETEST", BigDecimal("78.93")))
            it.apply(PaymentMade(BigDecimal("100.25")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)

            it.apply(ProcedurePerformed("ANNUALPHYSICAL", BigDecimal("170.00")))
            it.apply(PaymentMade(BigDecimal("180.00")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }
    }

    fun setupRingoStarr() {
        on(patientRepository.save(Patient("43"))) {
            it.apply(Created("Ringo Starr"))
            it.apply(ProcedurePerformed("ANNUALPHYSICAL", BigDecimal("170.00")))
            it.apply(ProcedurePerformed("GLUCOSETEST", BigDecimal("78.93")))
            it.apply(PaymentMade(BigDecimal("100.25")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)

            it.apply(ProcedurePerformed("FLUSHOT", BigDecimal("32.40")))
            it.apply(PaymentMade(BigDecimal("180.00")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }
    }

    fun setupPaulMcCartney() {
        on(patientRepository.save(Patient("44"))) {
            it.apply(Created("Paul McCartney"))
            it.apply(ProcedurePerformed("ANNUALPHYSICAL", BigDecimal("170.00")))

            val gluc = it.apply(ProcedurePerformed("GLUCOSETEST", BigDecimal("78.93")))

            it.apply(PaymentMade(BigDecimal("100.25")))
            it.apply(PatientEvent.Reverted(gluc.id!!))

            val pmt = it.apply(PaymentMade(BigDecimal("30.00")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)

            it.apply(PatientEvent.Reverted(pmt.id!!))
            it.apply(PaymentMade(BigDecimal("60.00")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)

            it.apply(PaymentMade(BigDecimal("60.00")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }
    }

    fun setupFreddieMercury() {
        val patient = patientRepository.save(Patient("45"))
        on(patient) {
            it.apply(Created("Farrokh Bulsara"))
            it.apply(ProcedurePerformed("ANNUALPHYSICAL", BigDecimal("170.00")))
            it.apply(ProcedurePerformed("GLUCOSE", BigDecimal("78.93")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }

        val patient2 = patientRepository.save(Patient("46"))
        on(patient2) {
            it.apply(Created("Freddie Mercury"))
            it.apply(PaymentMade(BigDecimal("100.25")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }

        currDate.add(Calendar.DATE, 1)
        merge(patient, patient2)
    }

    fun setupTinaFeyAndSarahPalin() {
        val patient = patientRepository.save(Patient("47"))
        on(patient) {
            it.apply(Created("Tina Fey"))
            it.apply(ProcedurePerformed("ANNUALPHYSICAL", BigDecimal("170.00")))
            it.apply(ProcedurePerformed("GLUCOSE", BigDecimal("78.93")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }

        val patient2 = patientRepository.save(Patient("48"))
        on(patient2) {
            it.apply(Created("Sarah Palin"))
            it.apply(PaymentMade(BigDecimal("100.25")))

            snapshotWith(it, patientAccountQuery, patientAccountRepository)
            snapshotWith(it, patientHealthQuery, patientHealthRepository)
        }

        currDate.add(Calendar.DATE, 1)
        val mergeEvent = merge(patient, patient2)

        on(patient) {
            it.apply(PatientEvent.Reverted(mergeEvent.id!!))
        }
        on(patient2) {
            it.apply(PatientEvent.Reverted(mergeEvent.converseId))
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun <
        SnapshotT : Snapshot<Patient, String, String, PatientEvent>,
        QueryT : QuerySupport<Patient, String, PatientEvent, String, SnapshotT>,
    > snapshotWith(
        it: OnSpec<
            Patient,
            String,
            PatientEvent,
            String,
            out Snapshot<Patient, String, String, PatientEvent>,
        >,
        query: QueryT,
        repository: ReactiveCrudRepository<SnapshotT, String>,
    ) = null
//            query.computeSnapshot(it.aggregate, Long.MAX_VALUE)
//                    .flatMap { repository.save(it).toObservable() }
//                    .toBlocking()
//                    .subscribe()

    /**
     *
     * @param self The aggregate to be deprecated
     * @param into The aggregate to survive
     * @return
     */
    private fun merge(
        self: Patient,
        into: Patient,
    ): PatientEvent.PatientDeprecatedBy {
        val e2 =
            patientEventRepository.save(
                PatientEvent.PatientDeprecates(self).also {
                    it.aggregate = into
                    it.timestamp = currDate.time
                    it.position = countEvents(into)
                },
            )

        val e1 =
            patientEventRepository.save(
                PatientEvent.PatientDeprecatedBy(into, e2.id!!).also {
                    it.aggregate = self
                    it.timestamp = currDate.time
                    it.position = countEvents(self)
                },
            )

        e2.converseId = e1.id

        patientEventRepository.save(e2)

        return e1
    }

    val currDate =
        Calendar
            .getInstance(TimeZone.getTimeZone("UTC"))
            .also {
                it.time =
                    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX")
                        .parse("2016-01-01T00:00:00.000Z")
            }

    fun on(
        patient: Patient,
        closure: (
            OnSpec<Patient, String, PatientEvent, String, out Snapshot<Patient, String, String, PatientEvent>>,
        ) -> Unit,
    ): Patient {
        val eventSaver: (Any) -> Unit = {
            when (it) {
                is PatientEvent -> patientEventRepository.save(it)
            }
        }
        val positionSupplier = { countEvents(patient) }
        val timestampSupplier = {
            currDate.add(Calendar.DATE, 1)
            currDate.time
        }
        return com.github.rahulsom.grooves.test
            .EventsDsl<Patient, String, PatientEvent>()
            .on<String, Snapshot<Patient, String, String, PatientEvent>>(
                patient,
                eventSaver,
                positionSupplier,
                timestampSupplier,
                closure,
            )
    }

    private fun countEvents(patient: Patient) =
        patientEventRepository
            .findAllByAggregateIdIn(listOf(patient.id!!))
            .count() + 1L
}